tests/test_common.py

```diff
@@ -265,34 +273,151 @@ def get(self, key: str, range_end: str = "", rev: int = 0):
+            self.check_receipt("put", req, res_pb)
+        return res
+
+    def delete(
+        self,
+        key: str,
+        range_end: str = "",
+        wait_for_commit: bool = True,
+        check_receipt=True,
+    ):
         """
         Perform a delete operation on lskv.
         """
         logger.info("Delete: {} {}", key, range_end)
-        j = {"key": b64encode(key)}
+        req = etcd_pb2.DeleteRangeRequest()
+        req.key = key.encode("utf-8")
         if range_end:
-            j["range_end"] = b64encode(range_end)
-        return self.client.post("/v3/kv/delete_range", json=j)
+            req.range_end = range_end.encode("utf-8")
+        j = MessageToDict(req)
+        res = self.client.post("/v3/kv/delete_range", json=j)
+        check_response(res)
+        if wait_for_commit:
+            rev, term = extract_rev_term(res)
+            self.wait_for_commit(term, rev)
+        if check_receipt:
+            res_pb = ParseDict(res.json(), etcd_pb2.DeleteRangeResponse())
+            self.check_receipt("delete_range", req, res_pb)
+        return res
+
+    def get_receipt(self, rev: int, term: int):
+        """
+        Get a receipt for a revision and term.
+        """
+        logger.info("GetReceipt: {} {}", rev, term)
+        req = lskvserver_pb2.GetReceiptRequest()
+        req.revision = rev
+        req.raft_term = term
+        j = MessageToDict(req)
+        res = self.client.post(
+            "/v3/receipt/get_receipt",
+            json=j,
+        )
+        if res.status_code == http.HTTPStatus.ACCEPTED:
+            logger.info("GetReceipt: ACCEPTED")
+            # accepted, retry
+            res = self.client.post(
+                "/v3/receipt/get_receipt",
+                json=j,
+            )
+        return res
 
     def raw(self) -> httpx.Client:
         """
         Get the raw client.
         """
         return self.client
+
+    def check_receipt(self, req_type: str, request, response):
+        """
+        Check a receipt for a request and response.
+        """
+        rev, term = extract_rev_term_pb(response)
+        res = self.get_receipt(rev, term)
+        check_response(res)
+        body = res.json()
+        assert "receipt" in body
+        receipt = body["receipt"]
+        assert "txReceipt" in receipt
+        tx_receipt = receipt["txReceipt"]
+        assert "leafComponents" in tx_receipt
+        leaf_components = tx_receipt["leafComponents"]
+        assert "claimsDigest" in leaf_components
+        claims_digest = leaf_components["claimsDigest"]
+
+        response.ClearField("header")
+
+        claims = lskvserver_pb2.ReceiptClaims()
+        getattr(claims, f"request_{req_type}").CopyFrom(request)
+        getattr(claims, f"response_{req_type}").CopyFrom(response)
+        claims_ser = claims.SerializeToString()
+        claims_digest_calculated = hashlib.sha256(claims_ser).hexdigest()
+        assert claims_digest == claims_digest_calculated
```

Comment:
I would have expected this function to make use of https://microsoft.github.io/CCF/main/audit/python_library.html#ccf.receipt.root and existing `ccf` receipt helper functions?

Reply:
Hmm, @jumaffre I'm getting errors after changing to use that (https://github.com/microsoft/LSKV/actions/runs/3446163543/jobs/5750731840)

I agree we should be using it though so would like to get this fixed.

Reply:
Apologies for the late reply. This isn't clear to me why the signature verification is failing. My bet would be that the calculated root isn't correct. If you have a local run, you could have a look in the ledger at the root in the [signatures](https://microsoft.github.io/CCF/main/audit/builtin_maps.html#signatures) entry that follows the transaction you get a receipt for and check that the root is valid there?

Reply:
Think I've fixed it, was hashing the hex representation rather than the bytes to get the root.